### PR TITLE
WFFHCOHORT-192: Adding an example flink driver

### PR DESCRIPTION
**Pull request by dkwasny-ibm**
Not a whole lot to say here.
There's a few optimizations and changes in the core projects in addition to the new submodule.

### CODE REVIEW

**Review by jillrdoty — APPROVED** [review]

**Review by csandersdev — APPROVED** [review]
LGTM

File: cohort-engine/src/main/java/com/ibm/cohort/engine/measure/ProviderFactory.java [hunk 1]
```diff
@@ -26,6 +26,12 @@
  */
 public class ProviderFactory implements EvaluationProviderFactory {
 
+	/*
+	 * Wrap the ModelResolver around a static ThreadLocal to prevent
+	 * excess creation of FhirContext instances.
+	 */
+	private static final ThreadLocal<ModelResolver> MODEL_RESOLVER = ThreadLocal.withInitial(R4FhirModelResolver::new);
```
Comment (deekim) [inline cohort-engine/src/main/java/com/ibm/cohort/engine/measure/ProviderFactory.java:33]: neat!

File: cohort-engine/src/main/java/com/ibm/cohort/engine/InsecureApacheRestfulClientFactory.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+	}
+
+	public HttpClient getNativeHttpClient() {
+		if (myHttpClient == null) {
+
+			PoolingHttpClientConnectionManager connectionManager = createInsecureHttpConnectionManager(5000, TimeUnit.MILLISECONDS);
+			connectionManager.setMaxTotal(getPoolMaxTotal());
+			connectionManager.setDefaultMaxPerRoute(getPoolMaxPerRoute());
+
+			//TODO: Use of a deprecated method should be resolved.
```
Comment (deekim) [inline cohort-engine/src/main/java/com/ibm/cohort/engine/InsecureApacheRestfulClientFactory.java:99]: stray `todo`

File: cohort-engine/src/main/java/com/ibm/cohort/engine/InsecureApacheRestfulClientFactory.java [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+	public void setProxy(String theHost, Integer thePort) {
+		if (theHost != null) {
+			myProxy = new HttpHost(theHost, thePort, "http");
+		} else {
+			myProxy = null;
+		}
+	}
+
+	private PoolingHttpClientConnectionManager createInsecureHttpConnectionManager(int timeToLive, TimeUnit timeUnit) {
+		X509TrustManager trustManager = new X509TrustManager() {
```
Comment (deekim) [inline cohort-engine/src/main/java/com/ibm/cohort/engine/InsecureApacheRestfulClientFactory.java:151]: this seems `instance`-able, or do you need a new one every time for thread stuff?

File: cohort-engine-flink-example/src/main/java/com/ibm/cohort/engine/flink/execution/CohortEngineFlinkDriver.java [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+				KafkaCommon.getConsumerProperties(kafkaInputInfo, kafkaGroupId)
+		);
+		if (readFromStart) {
+			consumer.setStartFromEarliest();
+		}
+
+		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+		DataStream<String> stream = env.addSource(consumer);
+		if (rebalanceInput) {
+			// TODO: Remove `setParallelism()` once our cluster is actually configured right.
```
Comment (deekim) [inline cohort-engine-flink-example/src/main/java/com/ibm/cohort/engine/flink/execution/CohortEngineFlinkDriver.java:110]: stray `todo`

File: cohort-engine-flink-example/src/main/java/com/ibm/cohort/engine/flink/execution/CohortEngineFlinkDriver.java [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+			DataStreamSource<String> cast = (DataStreamSource<String>)stream;
+			stream = cast.setParallelism(1).rebalance();
```
Comment (deekim) [inline cohort-engine-flink-example/src/main/java/com/ibm/cohort/engine/flink/execution/CohortEngineFlinkDriver.java:112]: just wanted to double-check we wanted to have this be `1` in version-control

**Review by deekim — APPROVED** [review]

File: cohort-engine/src/main/java/com/ibm/cohort/engine/InsecureApacheRestfulClientFactory.java (hunk 2 above)
Comment (dkwasny-ibm) [reply cohort-engine/src/main/java/com/ibm/cohort/engine/InsecureApacheRestfulClientFactory.java:99]: That todo is actually copied from [HAPI FHIR](https://github.com/hapifhir/hapi-fhir/blob/cbb16ce3affd3fc53dcbfe98dd3181644fe68604/hapi-fhir-client/src/main/java/ca/uhn/fhir/rest/client/apache/ApacheRestfulClientFactory.java#L93).

File: cohort-engine/src/main/java/com/ibm/cohort/engine/InsecureApacheRestfulClientFactory.java (hunk 3 above)
Comment (dkwasny-ibm) [reply cohort-engine/src/main/java/com/ibm/cohort/engine/InsecureApacheRestfulClientFactory.java:151]: That should be doable.

File: cohort-engine-flink-example/src/main/java/com/ibm/cohort/engine/flink/execution/CohortEngineFlinkDriver.java (hunk 5 above)
Comment (dkwasny-ibm) [reply cohort-engine-flink-example/src/main/java/com/ibm/cohort/engine/flink/execution/CohortEngineFlinkDriver.java:112]: My goal is to remove the entire `setParallelism()` call, but then the job will always fail in our cluster while the one K8s node can't communicate with Event Streams.  I plan to wait until ops fixes that before merging down.

File: cohort-engine-flink-example/src/main/java/com/ibm/cohort/engine/flink/execution/CohortEngineFlinkDriver.java (hunk 4 above)
Comment (dkwasny-ibm) [reply cohort-engine-flink-example/src/main/java/com/ibm/cohort/engine/flink/execution/CohortEngineFlinkDriver.java:110]: See your other comment.

**Review by msargentibm — APPROVED** [review]

**Review by tabishop — APPROVED** [review]
In general, more comments in the code would be helpful for those new to the code or unfamiliar with this area of the code.

**Pull request merged**